client/src/chat/Chat.jsx

```diff
@@ -0,0 +1,178 @@
+      id: actionId,
+      type: "send_message",
+      content: text,
+      targetId: undefined,
+      playerPosition: currentPlayerPosition,
+      sender: {
+        id: player?.id,
+        name: player?.get("name") || player?.id,
+        title: player?.get("title"),
+        avatar: player?.get("avatar"),
+      },
+      stage: progressLabel,
+      time: elapsed,
+    });
+  };
+
+  const handleAddReaction = (messageId, emoji) => {
+    // Check if reactions are enabled
+    if (!reactionEmojisAvailable || reactionEmojisAvailable.length === 0) {
+      return;
+    }
+
+    // Check numReactionsPerMessage limit
+    const message = messages.find((m) => m.id === messageId);
+    if (!message) return;
+
+    const currentPlayerReactions = message.reactions.filter(
+      (r) => r.playerPosition === currentPlayerPosition
+    );
+
+    if (currentPlayerReactions.length >= numReactionsPerMessage) {
+      // User has reached their limit, don't add more
+      return;
+    }
+
+    // Check if user already reacted with this specific emoji
+    const hasThisEmoji = currentPlayerReactions.some(
+      (r) => r.emoji === emoji
+    );
+    if (hasThisEmoji) {
+      // User already used this emoji, don't add duplicate
+      return;
+    }
+
+    const actions = scope.getAttribute(attribute)?.items || [];
+    const actionId = getNextActionId(actions);
+
+    scope.append(attribute, {
+      id: actionId,
+      type: "add_reaction_emoji",
+      content: emoji,
+      targetId: messageId,
+      playerPosition: currentPlayerPosition,
+      stage: progressLabel,
+      time: elapsed,
+    });
+  };
+
+  const handleRemoveReaction = (reactionId) => {
+    const actions = scope.getAttribute(attribute)?.items || [];
+    const actionId = getNextActionId(actions);
+
+    scope.append(attribute, {
+      id: actionId,
+      type: "remove_reaction_emoji",
+      content: undefined,
+      targetId: reactionId,
+      playerPosition: currentPlayerPosition,
+      stage: progressLabel,
+      time: elapsed,
+    });
+  };
+
+  if (messages.length === 0) {
+    return (
+      <div className="h-full w-full flex flex-col">
+        <div className="h-full w-full flex justify-center items-center">
+          <div className="flex flex-col justify-center items-center w-2/3 space-y-2">
+            <div className="w-24 h-24 text-gray-200">
+              <ChatIcon />
+            </div>
+
+            <h4 className="text-gray-700 font-semibold">No chat yet</h4>
+
+            <p className="text-gray-500 text-center">
+              Send a message to start the conversation.
+            </p>
+          </div>
+        </div>
+        <TextBar
+          onSendMessage={handleSendMessage}
+          reactionEmojisAvailable={reactionEmojisAvailable}
+        />
+      </div>
+    );
+  }
+
+  return (
+    <div className="h-full w-full flex flex-col">
+      <div className="h-full overflow-auto flex flex-col justify-end" ref={scroller}>
```

Comment:
actually, it looks like i can't scroll back in the chat history any more at all, when the messages are too many and overflow the top of the screen (as the new ones show up at the bottom first).